TOY-69 커스텀 갤러리화면 MVVM 적용

pby2017 opened the pull request:
코드 리뷰 부탁드립니다.

기존 코드를 MVVM 구조로 변경하면서 수정한 사항은 아래와 같습니다.

#### 1. loadImages(), getFolderNames() 함수를 Utils 파일로 이동했습니다.

#### 2. 데이터바인딩을 사용했습니다.
-> 적용 및 편집 버튼의 활성화(clickable) 여부 판단
```kotlin
android:clickable="@{(viewModel.recentPosition &lt; 0) ? false : true}"
```

#### 3. 버튼 클릭 이벤트 처리 로직을 ViewModel로 옮겼습니다.
-> 로직에 대한 정의는 View에서 확장함수로 정의
```kotlin
startDoneButtonAction = {...}
```

#### 4. 기존 ViewHolder로 선택된 사진을 관리하는 방법을 adapterPosition으로 관리하도록 수정했습니다.
-> 사진 선택 후 스크롤로 화면을 벗어나면 선택된 사진이 바뀌는 문제 해결

#### 5. Fragment에서 리스트를 PublishSubject 로 받아 Adapter의 리스트를 갱신하도록 수정했습니다.
```kotlin
galleryViewModel.loadImagesPublishSubject
                        .subscribe {...}
```

#### 6. LoadImage 데이터 클래스에 selected 프로퍼티를 추가했습니다.
-> 사진 선택했을 경우 이 값을 통해 화면에 선택되었다는 표시를 보여줌

#### 7. 불필요한 strings 내 코드를 제거했습니다.

pby2017 committed:
TOY-69 커스텀 갤러리화면 MVVM 적용

bonchangkoo commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/gallery/GalleryActivity.kt:
```diff
@@ -14,23 +16,54 @@ import com.gun0912.tedpermission.PermissionListener
 
 class GalleryActivity : BaseActivity() {
 
+    private val galleryViewModel by lazy {
```
viewModel을 이렇게 길게 쓰는 형태로 하지 말고, fun GalleryViewModel.initViewModel( ) 이렇게 사용하는 편이 더 좋을 거 같아요.

pby2017 replied:
네,
피드백 반영하겠습니다.

bonchangkoo reviewed (commented):
코멘트 확인 부탁해요~~~

pby2017 committed:
TOY-69 피드백 반영

pby2017 commented:
> 코멘트 확인 부탁해요~~~

@bonchangkoo 님,
코멘트 확인 후 피드백 반영했습니다.

코드 리뷰 부탁드립니다.

hyejin830 commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/gallery/GalleryActivity.kt:
```diff
@@ -167,10 +94,50 @@ class GalleryActivity : BaseActivity() {
     private fun setupCheckPermission() {
         val permissionListener = object : PermissionListener {
             override fun onPermissionGranted() {
+                (DataBindingUtil.setContentView(this@GalleryActivity, R.layout.activity_gallery) as ActivityGalleryBinding)
```
이렇게 짧게 쓰는 형태도 좋지만, 다른 액티비티에서 사용하는 형태와 통일성이 있으면 더 좋을 것 같습니다. 그리고 개인적으로 가독성이 조금 떨어지는 것 같습니다만 제 의견입니다...!

pby2017 replied:
네,
피드백 반영해서 통일성 맞추겠습니다.

bonchangkoo reviewed (approved):
확인했읍니다.

hyejin830 commented on app/src/main/java/kr/co/yogiyo/rookiephotoapp/gallery/GalleryViewModel.kt:
```diff
@@ -0,0 +1,53 @@
+
+    private var recentImagePath: String? = null
+
+    fun onClickDoneButton() {
```
on으로 시작하는 함수명은 콜백함수인 것 같으니..이름을 바꾸시는게 좋을 듯 합니다!

pby2017 replied:
네,
피드백 반영해서 아래처럼 수정하겠습니다.
```kotlin
// 적용버튼
onClickDoneButton -> startDoneButtonAction
startDoneButtonAction-> returnResult

// 편집버튼
onClickEditButton -> startEditButtonAction
startEditButtonAction -> sendResult
```

hyejin830 commented:
리팩토링 이후에 갤러리에서 사진 선택하는 속도가 좀 느려진 것 같습니다. 체크 부탁드립니다!

hyejin830 commented:
확인 완료했습니다!

pby2017 commented:
> 리팩토링 이후에 갤러리에서 사진 선택하는 속도가 좀 느려진 것 같습니다. 체크 부탁드립니다!

@hyejin830 님,
코드 수정을 할 게 좀 있어보여서
이후 새로운 작업에서 피드백 반영하겠습니다.

pby2017 committed:
TOY-69 피드백 반영

pby2017 committed:
TOY-69 충돌해결 및 파일 위치 수정

Merge branch 'develop' into TOY-69-gallery-custom-mvvm

# Conflicts:
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/gallery/LoadImage.kt
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/ui/camera/capture/CameraActivity.kt
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/ui/camera/gallery/GalleryActivity.kt
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/ui/camera/gallery/GalleryFragment.kt
#	app/src/main/java/kr/co/yogiyo/rookiephotoapp/utils/ImageUtils.kt
#	app/src/main/res/layout/activity_gallery.xml
